NXCM-3600: exposed fix

cstamas opened the pull request:
CI job:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/129/

cstamas committed:
NXCM-3600: fixing content plexus resource to honor isExposed flag.

Since not all ResourceStoreContentPlexusResource extending classes are going
over Router, some of them go directly with Repository instances, and repo instance
itself does not care about isExposed. That flag is meant for _transport layer_
to expose it or not for external request.

cstamas committed:
NXCM-3600: adding comments about actual bug here

But unable to fix it from here. The problem is "piggybacked" method
that is not meant to be used like this...

cstamas committed:
NXCM-3600: always doing security based decisions

Added way back in commit 71b6a9d57c99914c5e1e3b7fbe6295e9187eb86b

The reasons for doing this was _probably_ to enable internal tasks
(unsolved back than, had no security context) to still run and work
with unpublished repositories. Hoping the best.

bdemers commented on nexus/nexus-rest-api/src/main/java/org/sonatype/nexus/rest/AbstractResourceStoreContentPlexusResource.java:
```diff
@@ -252,7 +255,35 @@ protected String parsePathFromUri( String parsedPath )
+        {
+            if ( !( (Repository) resourceStore ).isExposed() )
+            {
+                throw new NoSuchRepositoryException( ( (Repository) resourceStore ).getId() );
```
Not completely true, as the repo does exist. Though, I think this will do, as it is almost true.

cstamas replied:
Very same things happen with "protected" repository registry and user not having "view" perm. For "surface", both of those end up with 404 Not found.

bdemers commented on nexus/nexus-rest-api/src/main/java/org/sonatype/nexus/rest/AbstractResourceStoreContentPlexusResource.java:
```diff
@@ -252,7 +255,35 @@ protected String parsePathFromUri( String parsedPath )
      * @throws NoSuchRepositoryRouterException
      */
-    protected abstract ResourceStore getResourceStore( Request request )
+    protected ResourceStore getResourceStore( final Request request, final Action action )
```
Verify this method is not overridden in a PRO plugin (we need to know what to back port 1.9.2.4)

cstamas replied:
Read the diff better, the original method is intact, it is just a new method is added with different signature (but the diff probably mislead you), and it is called from abstract class instead of the abstract getResourceStore( req ).

Subclasses needs no change, unless they override one of the get(), upload() or delete() methods. In core, there is only the RRB plugins class. Will see what's in Pro.

bdemers replied:
I see it now

bdemers commented:
Changes look good, but add a test ;)

cstamas commented:
@bdemers Sure, IT on the way, I just wanted to present needed changes for you to review them, since they are related to introduction of "view" perm made by you.

cstamas commented:
CI likes it:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/129/

cstamas committed:
NXCM-3600: Added IT, but not quite.

The IT does validates partially the assumptions in issue (that are fixed,
that is), but not quite: as I read the code, disabling anon access is not the same
as making a repo private without disabling anon access (see [1]). So, this
IT would still need some refinements here....

[1] https://docs.sonatype.com/display/SPRTNXOSS/Nexus+FAQ#NexusFAQ-Q.CanImakearepositoryprivatewithoutdisablinganonymousaccess%3F

cstamas committed:
NXCM-3600: Adding missing header.

cstamas commented:
CI likes it even more:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/130/

Note: while both slaves failed the build (!), there was no UT/IT error, but it was the fact that very last "licence check" failed, since newly added IT had no license header.

This is already fixed in last commit (e2eca1a).

cstamas committed:
NXCM-3600: Reworking ITs a bit, added 2nd IT

2nd IT covers the "real thing", it uses a "private repository" and does basically the
same access tests as 1st one, but notice the slight differences in response codes!

cstamas commented:
Added 2nd IT too.

bdemers commented on nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nxcm3600/AbstractNxcm3600IntegrationTest.java:
```diff
@@ -0,0 +1,97 @@
+        throws IOException
+    {
+        TestContainer.getInstance().getTestContext().useAdminForRequests();
+        TestContainer.getInstance().getTestContext().setSecureTest( true );
```
setSecureTest should already be called called by the parent class

I see now... sorry

bdemers commented:
+1
though, I'd like to see a PUT along with the GET and DELETE tests

cstamas commented:
One more CI +1:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/132/

nabcos commented:
+1
